tests/Microsoft.Identity.Test.Unit.net45/CacheTests/TokenCacheNotificationTests.cs

```diff
@@ -0,0 +1,148 @@
+
+            bool beforeAccessCalled = false;
+            bool afterAccessCalled = false;
+            bool beforeWriteCalled = false;
+
+            bool asyncBeforeAccessCalled = false;
+            bool asyncAfterAccessCalled = false;
+            bool asyncBeforeWriteCalled = false;
+
+            // Sync method should be called _first_ (just by convention).  But let's validate this.
+
+            pca.UserTokenCache.SetBeforeAccess(args => { beforeAccessCalled = true; });
+            pca.UserTokenCache.SetAfterAccess(args => { afterAccessCalled = true; });
+            pca.UserTokenCache.SetBeforeWrite(args => { beforeWriteCalled = true; });
+
+            pca.UserTokenCache.SetBeforeAccessAsync(async args => { asyncBeforeAccessCalled = beforeAccessCalled; await Task.Delay(10).ConfigureAwait(false); });
+            pca.UserTokenCache.SetAfterAccessAsync(async args => { asyncAfterAccessCalled = afterAccessCalled; await Task.Delay(10).ConfigureAwait(false); });
+            pca.UserTokenCache.SetBeforeWriteAsync(async args => { asyncBeforeWriteCalled = beforeWriteCalled; await Task.Delay(10).ConfigureAwait(false); });
+
+            await pca.GetAccountsAsync().ConfigureAwait(false);
+
+            Assert.IsTrue(asyncBeforeAccessCalled);
+            Assert.IsTrue(asyncAfterAccessCalled);
+            Assert.IsFalse(asyncBeforeWriteCalled);
+
+            Assert.IsTrue(beforeAccessCalled);
+            Assert.IsTrue(afterAccessCalled);
+            Assert.IsFalse(beforeWriteCalled);
+        }
+
+        [TestMethod]
+        public async Task TestSerializationViaAsync()
+        {
+            int numBeforeAccessCalls = 0;
+            int numAfterAccessCalls = 0;
+            int numBeforeWriteCalls = 0;
+
+            byte[] serializedPayload = null;
+
+            using (var harness = new MockHttpAndServiceBundle())
+            {
+                harness.HttpManager.AddInstanceDiscoveryMockHandler();
+
+                PublicClientApplication pca = PublicClientApplicationBuilder
+                    .Create(MsalTestConstants.ClientId)
+                    .WithAuthority(new Uri(ClientApplicationBase.DefaultAuthority), true)
+                    .WithHttpManager(harness.HttpManager)
+                    .BuildConcrete();
+
+                pca.UserTokenCache.SetBeforeAccessAsync(async args =>
+                {
+                    numBeforeAccessCalls++;
+                    await Task.Delay(10).ConfigureAwait(false);
+                });
+                pca.UserTokenCache.SetAfterAccessAsync(async args =>
+                {
+                    numAfterAccessCalls++;
+                    serializedPayload = args.TokenCache.SerializeMsalV3();
+                    await Task.Delay(10).ConfigureAwait(false);
+                });
+                pca.UserTokenCache.SetBeforeWriteAsync(async args =>
+                {
+                    numBeforeWriteCalls++;                    
```

Comment:
I wonder if there a risk that some of the async calls are not syncronized. We could strengthen this test by checking not only if the callbacks are hit, but also the order in which they are hit. Smth like append "beforeAccess-", "afterAccess-" etc. to a stringbuilder and then check that the final string is "beforeAccess-afterAccess-afterWrite" etc.

Reply:
good suggestion.  fixed.